Add IPV6 support to built-in server for run-benchmarks

jcarterbohan opened the pull request:
#### ecc5b337e0f999d49c89eba273e15483d9231015
<pre>
Add IPV6 support to built-in server for run-benchmarks
<a href="https://bugs.webkit.org/show_bug.cgi?id=265402">https://bugs.webkit.org/show_bug.cgi?id=265402</a>

Reviewed by Dewei Zhu.

Using socket.getaddrinfo() allows us to determine if the address is IPv4 or IPv6, and we can then adjust the address family in HTTPServer accordingly.

* Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/builtin_http_server.py:

Canonical link: <a href="https://commits.webkit.org/271412@main">https://commits.webkit.org/271412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b34a72c439995ff3c22a5d5e76cfb5a0deb4eb3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6611 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29275 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30503 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25515 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28468 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8605 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4002 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25279 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28240 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5364 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24035 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4621 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/28321 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25042 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31192 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25576 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25478 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31060 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2972 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28871 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6343 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6779 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5253 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5294 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xw commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/builtin_http_server.py:
```diff
@@ -62,10 +63,13 @@ def do_POST(self):
     parser.add_argument('--log-path', default='/tmp/run-benchmark-http.log')
     args = parser.parse_args()
+    addr_info = socket.getaddrinfo(args.interface, args.port)
+    addr_pair = addr_info[0][4]
```
This should handle cases where `socket.getaddrinfo` errors out (usually through raising a `socket.gaierror`, though the error message isn't terribly informative), produces an empty list, or produces a list with more than one info tuple. 

Erroring out and empty list should be pretty easy to handle - we want to halt and raise our own (perhaps more informative) error message. But the question of which info tuple to select when multiple are provided is an interesting question - something more robust would be to always select IPV6 if it's available and otherwise fall back to IPV4. However, I can also see a case for just selecting the first one in the list, especially since in the context of the run-benchmarks harness we generally pass in IP addresses (`127.0.0.1`) and not a domain (`localhost`)

jcarterbohan replied:
I think we can use the first address returned by `socket.getaddrinfo` because from what `simple_http_server_driver` suggests, the interface should be passed in a IP address over a domain which should only return 1 option. Additionally this is the similar method which Twisted uses when handling an IPV6 address: https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/tcp.py#L649

If there is preference for IPV6 addresses then I think that could be saved for a future feature if domains are being passed in.

I agree that we should raise our own error messages on socket.gaierror exceptions.

dewei-zhu reviewed (approved):
r=me

jcarterbohan committed:
Add IPV6 support to built-in server for run-benchmarks
https://bugs.webkit.org/show_bug.cgi?id=265402

Reviewed by Dewei Zhu.

Using socket.getaddrinfo() allows us to determine if the address is IPv4 or IPv6, and we can then adjust the address family in HTTPServer accordingly.

* Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/builtin_http_server.py:

Canonical link: https://commits.webkit.org/271412@main
